WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift

```diff
@@ -797,6 +811,65 @@ extension ProductSelectorViewModel {
     }
 }
 
+// MARK: - Tracking
+private extension ProductSelectorViewModel {
+    func trackConfirmButtonTapped(with productsCount: Int) {
+        let trackingSources = Array(productIDTrackingSources.values.map { $0.rawValue })
+        let areFiltersActive = filtersSubject.value.numberOfActiveFilters > 0
+        analytics.track(event: WooAnalyticsEvent.Orders.orderCreationProductSelectorConfirmButtonTapped(productCount: productsCount,
+                                                                                                        sources: trackingSources,
+                                                                                                        isFilterActive: areFiltersActive))
+    }
+
+    func updateTrackingSourceAfterSelectionStateChangedForProduct(with productID: Int64) {
+        guard productIDTrackingSources[productID] == nil else {
+            productIDTrackingSources.removeValue(forKey: productID)
+
+            return
+        }
+
+        if let trackingSource = retrieveTrackingSource(for: productID) {
+            productIDTrackingSources[productID] = trackingSource
+        }
+    }
+
+    func updateTrackingSourceAfterSelectionStateChangedForProduct(with productID: Int64, selectedVariationIDs: [Int64]) {
+        guard selectedVariationIDs.isNotEmpty else {
+            productIDTrackingSources.removeValue(forKey: productID)
+
+            return
+        }
+
+        productIDTrackingSources[productID] =  retrieveTrackingSource(for: productID)
+    }
+
+    func retrieveTrackingSource(for productID: Int64) -> ProductTrackingSource? {
+        guard topProductsFromCachedOrders != .empty else {
+            return nil
+        }
+
+        guard searchTerm.isEmpty else {
+            return .search
+        }
+
+        guard !sectionContainsProductID(sectionType: .mostPopular, productID: productID) else {
+            return .popular
+        }
+
+        guard !sectionContainsProductID(sectionType: .lastSold, productID: productID) else {
+            return .recent
+        }
+
+        return .alphabetical
+    }
+
+    func sectionContainsProductID(sectionType: ProductsSectionType, productID: Int64) -> Bool {
+        let section = sections.first(where: { $0.type == sectionType })
+
+        return section?.products.first(where: { $0.productID == productID}) != nil
+    }
+}
+
```

Comment:
I'm just wondering here: As all those methods hold the tracking responsibility, would it be worth to make a separate object that just hold static methods? For example: 

```
final class ProductSelectorViewModelTracker {
    private init() {}
    static func trackConfirmButtonTapped(with productCount: Int) {
     // Implementation...
    }

    // Rest of methods...
}
```

Then call it directly:

```diff
 func completeMultipleSelection() {
        let allIDs = selectedProductIDs + selectedProductVariationIDs
-      trackConfirmButtonTapped(with: allIDs.count)
+      ProductSelectorViewModelTracker.trackConfirmButtonTapped(with: allIDs.count)
        onMultipleSelectionCompleted?(allIDs)
    }

```

I'm thinking this could help reduce the view model size and separate more clearly where these belong. Perhaps move there `ProductTrackingSource` as well. Thoughts? 🤔

Reply:
Yeah, good idea, done in https://github.com/woocommerce/woocommerce-ios/pull/9586